uaclient/actions.py

```diff
@@ -67,6 +68,16 @@ def attach_with_token(
     )
 
     cfg.machine_token_file.write(new_machine_token)
+    ret, entitlement_clash_info = check_entitlement_directives_are_unique(cfg)
+
+    if not ret and entitlement_clash_info:
+        cfg.machine_token_file.delete()
+        raise exceptions.EntitlementsDirectivesAreNotUnique(
+            url=cfg.contract_url,
+            names=", ".join(sorted(entitlement_clash_info.entitlements)),
+            apt_url=entitlement_clash_info.apt_url,
+            suite=entitlement_clash_info.suite,
+        )
```

Comment:
We can move this around a bit to avoid a write-then-immediately-delete case.
```suggestion
    ret, entitlement_clash_info = check_entitlement_directives_are_unique(cfg)
    if not ret and entitlement_clash_info:
        raise exceptions.EntitlementsDirectivesAreNotUnique(
            url=cfg.contract_url,
            names=", ".join(sorted(entitlement_clash_info.entitlements)),
            apt_url=entitlement_clash_info.apt_url,
            suite=entitlement_clash_info.suite,
        )
        
    cfg.machine_token_file.write(new_machine_token)
```

Reply:
Ah sorry, is there any reason that that `check_entitlement_directives_are_unique` function cannot operate directly on `new_machine_token` here? So the call would become:
```python
    ret, entitlement_clash_info = check_entitlement_directives_are_unique(new_machine_token)
```

Reply:
Oh I see. The logic relies on creating the entitlement instances, which take the `cfg` and read the machine token that way.

This is okay then, no need to follow my suggestion.